### PR TITLE
Fix `aten::index_put` args Dtensor type mismatch and add a propagation strategy

### PR DESCRIPTION
**Pull request by zpcore**
We notice model code contains indexing syntax like [nanogpt model code](https://github.com/pytorch/benchmark/blob/f144fe9095291454f1e6a05d786f1aaad93a83d8/torchbenchmark/models/nanogpt/model.py#L240), which causes training fail in the backward pass when using DTensor.

In the code, `x = x[:, [-1], :]` calls the index op and in the backward pass, it will trigger `aten.index_put.default` with the second argument to be of type `torch::List<std::optional<Tensor>>`, e.g., `[None, tensor([-1], device='cuda:0')]`. We are unable to unwarp the op info into Dtensor based on the current logic [here](https://github.com/pytorch/pytorch/blob/2625c70aecc6eced1dbe108279feab7509733bef/torch/distributed/tensor/_dispatch.py#L339-L358). We need to set runtime_schema_info for the op and enable needs_pytree to support the conversion of tensor list arg.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @tianyu-l @XilunWu

### CODE REVIEW

File: test/distributed/tensor/test_tensor_ops.py [hunk 1]
```diff
@@ -440,7 +440,7 @@ def test_gather(self):
 
         # case 1 all replicate: input replicated, index replicated, output replicated
         global_input = torch.randn(12, 8, 16)
-        global_index = torch.randint(8, (4, 4, 8))
+        global_index = torch.randint(self.world_size, (4, 4, 8))
```
Comment (XilunWu) [inline test/distributed/tensor/test_tensor_ops.py]: why do we need this change?

File: test/distributed/tensor/test_tensor_ops.py [hunk 2]
```diff
@@ -592,6 +592,19 @@ def test_index(self):
                 torch.randint(5, (12, 8, 12)),
             )
 
+    @with_comms
+    def test_index_put(self):
+        device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
+        global_input = torch.randn(2, 2)
```
Comment (XilunWu) [inline test/distributed/tensor/test_tensor_ops.py]: let's use the `init_device_mesh` API instead of using `DeviceMesh` constructor.

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 3]
```diff
@@ -68,6 +68,16 @@ def default_strategy(op_schema: OpSchema) -> StrategyType:
...
 
+
+register_op_strategy(
+    [
+        aten.index_put.default,
+        aten._index_put_impl_.default,
+    ],
+    schema_info=RuntimeSchemaInfo(needs_pytree=True),
+)(default_strategy)
+
```
Comment (XilunWu) [inline torch/distributed/tensor/_ops/_tensor_ops.py]: We need to write down that the PR only enables the `Replicate` sharding for input DTensors on `index_put`.

File: test/distributed/tensor/test_tensor_ops.py [hunk 4]
```diff
@@ -592,6 +592,19 @@ def test_index(self):
...
+        global_index = [
+            i.detach().clone()
+            for i in torch.randperm(global_input.ndim)[: min(global_input.ndim, 2)]
```
Comment (XilunWu) [inline test/distributed/tensor/test_tensor_ops.py]: why do we need `min(global_input.ndim, 2)` since `global_input.ndim == 2`?

File: test/distributed/tensor/test_tensor_ops.py (hunk 2 above)
Comment (XilunWu) [inline test/distributed/tensor/test_tensor_ops.py]: let's add `device=self.device_type` for all tensor constructors.

File: test/distributed/tensor/test_tensor_ops.py [hunk 5]
```diff
@@ -592,6 +592,19 @@ def test_index(self):
...
+        ]
+        input_dt = distribute_tensor(global_input, device_mesh, [Replicate()])
+        output_dt = torch.index_put(input_dt, global_index, torch.tensor(1.0))
+        ref = torch.index_put(global_input, global_index, torch.tensor(1.0))
+        self.assertEqual(output_dt.full_tensor(), ref)
```
Comment (XilunWu) [inline test/distributed/tensor/test_tensor_ops.py:635]: torch.tensor(1.0, device=self.device_type)

**Review by XilunWu — COMMENTED** [review]
We need to rewrite the `index_put` sharding strategy. Let's start with always resharding the input to `Replicate` and see if there's anything can be optimized.

File: test/distributed/tensor/test_tensor_ops.py (hunk 1 above)
Comment (zpcore) [reply test/distributed/tensor/test_tensor_ops.py]: I thought the number should be limited to world_size but seems like this is not the case. I have reverted this change.

File: test/distributed/tensor/test_tensor_ops.py (hunk 4 above)
Comment (zpcore) [reply test/distributed/tensor/test_tensor_ops.py]: Right, since `global_input.ndim` is fixed, we don't need that. Removed.

File: torch/distributed/tensor/_ops/_tensor_ops.py (hunk 3 above)
Comment (zpcore) [reply torch/distributed/tensor/_ops/_tensor_ops.py]: My understanding is that `default_strategy` is using the sharding strategy based on args[0]: https://github.com/pytorch/pytorch/blob/12b02137af5cd6743adfe5a0ffe0d9b453cd013e/torch/distributed/tensor/_ops/_tensor_ops.py#L43. However, if I update the `input_dt` in the test to different placement, it will error out. `Replicate` is the only support now.

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 6]
```diff
@@ -468,6 +469,13 @@ def gen_slice_scatter_strategy(op_schema: OpSchema) -> StrategyType:
 
 
 @register_op_strategy(aten._local_scalar_dense.default)
+@register_op_strategy(
+    [
+        aten.index_put.default,
+        aten._index_put_impl_.default,
```
Comment (wanchaol) [inline torch/distributed/tensor/_ops/_tensor_ops.py]: hmmm I don't think index_put should use the replica_only_strategy, it should be similar to the [`index_select` strategy](https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_ops/_tensor_ops.py#L701).

**Review by wanchaol — COMMENTED** [review]
Please see the inlined comments. IMO we should not add arbitrary op to the replica_only strategy, if this PR intends to add the operator support for `index_put`. If one want to optimize the op coverage, one thing a person can do is to add all ops to this `replica_only_strategy` and claim the op coverage is up, but it does not really increase the op coverage in a meaningful way.

File: torch/distributed/tensor/_ops/_tensor_ops.py (hunk 6 above)
Comment (zpcore) [reply torch/distributed/tensor/_ops/_tensor_ops.py]: Oh, thanks for pointing out the strategy!

Comment (zpcore) [reply torch/distributed/tensor/_ops/_tensor_ops.py]: After taking a look at the `prop_index_select` strategy, I think `index_put` doesn't fit into the rule. Suppose there exist `Shard` placement for `index_put` input tensor, then in most cases only one sharding's element should be updated. If this is the case, we can't execute a uniform operation to all shards. I am afraid we have to default to `replica_only_strategy` at this moment even though it is not efficient.  wdyt?

Comment (wanchaol) [reply torch/distributed/tensor/_ops/_tensor_ops.py]: hmmm I looked at this again, isn't it have almost the same requirement as the `prop_index` (not the index_select one I guess) strategy? The index_put basically should allow: `input` be replicated, and `indices` and `values` have the same sharding

Comment (wanchaol) [reply torch/distributed/tensor/_ops/_tensor_ops.py]: Oh I see what's going on there, it's because the output is not well defined, as even if the indices is sharded and value is sharded, after the `index_put_`, the output becomes non-replicate anymore. Yeah for this op it seems we only have the choice of replicating everything

Comment (zpcore) [reply torch/distributed/tensor/_ops/_tensor_ops.py]: Hah, we both changed the opinion. Now I think `prop_index` rule can be applied for the following special case:
```
>>> torch.index_put(torch.zeros([2,4], dtype=torch.float), [torch.tensor(1)], torch.randn([4]))
tensor([[ 0.0000,  0.0000,  0.0000,  0.0000],
        [-0.5953,  0.7429,  0.0051, -0.3337]])
```
`torch.randn([4])` is the value to put into the origin tensor. `index_put` applies value to origin tensor's dim0 if the value is not a scalar as long as the shape matches. In this case we may reuse `prop_index`. We are unable to use `prop_index` if the last arg value is a scalar. Did I understand correctly? I didn't quite understand the `index_put_` case you mentioned above.

Comment (wanchaol) [reply torch/distributed/tensor/_ops/_tensor_ops.py]: > I didn't quite understand the index_put_ case you mentioned above.

The semantic of `index_put` is actually a bit confusing, the cases I mentioned above assumes the indices is a single tensor that semantically you can just use that single tensor to represent all the indices you want to put values on, and the values have the exact shape as the `indices` tensor. For this scenario the `prop_index` is almost the same strategy. But in this case, suppose `indices` is a sharded tensor, the `values` could be anything, `input` is replicated, after the `index_put`, each rank the `input` would only have certain entries be modified, so the `input` is neither sharded or replicated..

But `index_put_` documentation seems mentioning `indices` is a tuple of tensor, but does not explain the exact semantic, I think a way to interpret it is that the tuple of tensors possibly represent the indices of each tensor dimension. Even in this case, if `indices` is sharded, then the values must be sharded in the same way, but the results seems still can't be represented either as sharded or replicated. That's why I thought it might only have a replica_only_strategy for most of cases. For specific cases like you mentioned above, or other cases where the indices on the tensor dimension span the full tensor dimension (i.e. the input is [2,4] and indices is torch.arange(2), then if the indices is sharded, output is also sharded).

Let me know if that make sense :)

**Review by wanchaol — APPROVED** [review]
reasonable to keep replica_only first, feel free to add specialized cases that could propagate sharding if you like!

File: torch/distributed/tensor/_ops/_tensor_ops.py (hunk 6 above)
Comment (zpcore) [reply torch/distributed/tensor/_ops/_tensor_ops.py]: Thanks for the explanation! I have updated the strategy. I think we can make it simple to force `indices` to be replicated. I can hardly imagine how `value` can be placed into `input` when `indices` is sharded. I feel like our understanding of `indices` is different. Below are 3 examples on how indices is used to place different shape of `value`:
```
example 1: torch.index_put(torch.zeros([2,4,8], dtype=torch.float), [torch.tensor(1),torch.tensor(2),torch.tensor(3)], torch.randn(size=()))      <--- place a random value at indices [1,2,3]
example 2: torch.index_put(torch.zeros([2,4,8], dtype=torch.float), [torch.tensor(1),torch.tensor(2)], torch.randn([8]))    <--- place a random tensor of size [8] at indices [1,2,:]
example 3: torch.index_put(torch.zeros([2,4,8], dtype=torch.float), [torch.tensor(1)], torch.randn([4,8]))     <--- place a random tensor of size [4,8] at indices [1, :, :]
```
This is kind of different from `index_select`, where we pick up a dim and select chunks on top of the selected dim.

Here is the new strategy, basically 1) force `indices` to be replicated 2) allow `input` and `value` to be sharded on some axises but follows some criterias. Take example 3 as the case. We need to force `input` dim (dim=0 in example 3) that is selected to place `value` to be replicated, and other dims between `indices` (dim 1,2) and `values` (dim 0,1) to be the same sharding. That is to say,  `indices` dim 1 and `values` dim 0 use the same sharding (Shard()/Replicate()), `indices` dim 2 and `values` dim 1 sharded use the same sharding (Shard()/Replicate()). 

Hope I made it clear here.
